File: force-field/java/Particle.java
class Particle {
  
  public Vector position;
  public double mass;
  public Vector velocity;
  
  public Particle(Vector pos, double _mass) {
    position = pos;
    mass = _mass;
    velocity = new Vector();
  }
  
  public Vector forceFrom(Particle p) {
    Vector distance = p.position.subtract(position);
    double magnitude = mass * p.mass / (distance.dot(distance));
    return distance.normalize().scale(magnitude);
  }
  
  public void applyForce(Vector force, double timeDelta) {
    velocity.addInPlace(force.scale(timeDelta / Math.abs(mass)));
    position.addInPlace(velocity.scale(timeDelta));
  }
  
  public String toString() {
    return "{Particle mass: " + mass + " velocity: " + velocity +
      " position: " + position + "}";
  }
  
}

File: force-field/js/particle.js
var Vector = require('./vector');

function Particle(position, mass) {
  this.position = position;
  this.mass = mass;
  this.velocity = new Vector();
}

Particle.prototype.forceFrom = function(p) {
  var distance = p.position.subtract(this.position);
  var magnitude = this.mass * p.mass / distance.dot(distance);
  return distance.normalize().scale(magnitude);
};

Particle.prototype.applyForce = function(force, timeDelta) {
  this.velocity.addInPlace(force.scale(timeDelta / Math.abs(this.mass)));
  this.position.addInPlace(this.velocity.scale(timeDelta));
};

Particle.prototype.toString = function() {
  return '{Particle mass: ' + this.mass + ' velocity: ' + this.velocity
    + ' position: ' + this.position + '}';
};

module.exports = Particle;
